src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeReader.java

```diff
@@ -0,0 +1,455 @@
+        {
+            context.checkpoint();
+
+            final Relation r = visitor.compare(cellMinPacked, cellMaxPacked);
+
+            if (r == Relation.CELL_OUTSIDE_QUERY)
+            {
+                // This cell is fully outside the query shape: stop recursing
+                return;
+            }
+
+            if (r == Relation.CELL_INSIDE_QUERY)
+            {
+                // This cell is fully inside the query shape: recursively add all points in this cell without filtering
+                super.collectPostingLists(postingLists);
+                return;
+            }
+
+            if (index.isLeafNode())
+            {
+                if (index.nodeExists())
+                    filterLeaf(postingLists);
+                return;
+            }
+
+            visitNode(postingLists, cellMinPacked, cellMaxPacked);
+        }
+
+        private void filterLeaf(PriorityQueue<PeekablePostingList> postingLists) throws IOException
+        {
+            treeInput.seek(index.getLeafBlockFP());
+
+            int count = treeInput.readVInt();
+
+            // loading doc ids occurred here prior
+
+            int orderMapLength = treeInput.readVInt();
+
+            long orderMapPointer = treeInput.getFilePointer();
+
+            SeekingRandomAccessInput randomAccessInput = new SeekingRandomAccessInput(treeInput);
+            for (int x = 0; x < count; x++)
+            {
+                origIndex[x] = (short) LeafOrderMap.getValue(randomAccessInput, orderMapPointer, x, leafOrderMapReader);
+            }
+
+            // seek beyond the ordermap
+            treeInput.seek(orderMapPointer + orderMapLength);
+
+            FixedBitSet fixedBitSet = visitDocValues(treeInput, count, visitor, origIndex);
+
+            int nodeID = index.getNodeID();
+
+            if (postingsIndex.exists(nodeID) && fixedBitSet.cardinality() > 0)
+            {
+                long pointer = postingsIndex.getPostingsFilePointer(nodeID);
+                postingLists.add(initFilteringPostingReader(pointer, fixedBitSet));
+            }
+        }
+
+        void visitNode(PriorityQueue<PeekablePostingList> postingLists, byte[] cellMinPacked, byte[] cellMaxPacked) throws IOException
+        {
+            byte[] splitPackedValue = index.getSplitPackedValue();
+            BytesRef splitDimValue = index.getSplitDimValue();
+            assert splitDimValue.length == bytesPerValue;
+
+            // make sure cellMin <= splitValue <= cellMax:
+            assert FutureArrays.compareUnsigned(cellMinPacked, 0, bytesPerValue, splitDimValue.bytes, splitDimValue.offset, splitDimValue.offset + bytesPerValue) <= 0 : "bytesPerValue=" + bytesPerValue;
+            assert FutureArrays.compareUnsigned(cellMaxPacked, 0, bytesPerValue, splitDimValue.bytes, splitDimValue.offset, splitDimValue.offset + bytesPerValue) >= 0 : "bytesPerValue=" + bytesPerValue;
```

Comment:
nit: These checks feel expensive, but I suppose there would be real safety issues if the split value was out of range. I guess we should just make the messages more helpful.

Reply:
I would prefer to wrap these in the `DEBUG` flag we are using in the writer. This allows us to enable tree debugging during testing rather than rely on assertions which can be disabled.